File: apps/jump-android/app/src/main/java/com/youthperformance/jump/xlens/IMURecorder.kt
// ═══════════════════════════════════════════════════════════════
// IMU RECORDER
// High-frequency motion data capture for xLENS verification
// THE MOAT: AI can fake pixels, not synchronized G-forces
// ═══════════════════════════════════════════════════════════════

package com.youthperformance.jump.xlens

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Records accelerometer and gyroscope data at ~100Hz
 * This is the unfakeable signal that closes the "Analog Hole"
 */
class IMURecorder(context: Context) : SensorEventListener {

    companion object {
        private const val TAG = "IMURecorder"
        private const val TARGET_SAMPLE_RATE_HZ = 100.0
        private const val SAMPLE_DELAY_US = (1_000_000 / TARGET_SAMPLE_RATE_HZ).toInt()
    }

    // Sensor manager
    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
    private val gyroscope = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)

    // State
    private val _isRecording = MutableStateFlow(false)
    val isRecording: StateFlow<Boolean> = _isRecording.asStateFlow()

    private val _sampleCount = MutableStateFlow(0)
    val sampleCount: StateFlow<Int> = _sampleCount.asStateFlow()

    private val _currentAcceleration = MutableStateFlow(1.0)
    val currentAcceleration: StateFlow<Double> = _currentAcceleration.asStateFlow()

    // Recording data
    private var samples = mutableListOf<IMUSample>()
    private var startTimeMs: Long = 0

    // Latest sensor values
    private var lastAccelX = 0.0
    private var lastAccelY = 0.0
    private var lastAccelZ = 0.0
    private var lastGyroX = 0.0
    private var lastGyroY = 0.0
    private var lastGyroZ = 0.0
    private var hasAccelData = false
    private var hasGyroData = false

    // Device model
    private val deviceModel: String = "${Build.MANUFACTURER}_${Build.MODEL}"

    /**
     * Start recording IMU data
     */
    fun startRecording() {
        if (_isRecording.value) return

        if (accelerometer == null || gyroscope == null) {
            Log.e(TAG, "Motion sensors not available")
            return
        }

        // Reset state
        samples.clear()
        _sampleCount.value = 0
        startTimeMs = System.currentTimeMillis()
        _isRecording.value = true
        hasAccelData = false
        hasGyroData = false

        // Register listeners with high frequency
        sensorManager.registerListener(
            this,
            accelerometer,
            SAMPLE_DELAY_US
        )
        sensorManager.registerListener(
            this,
            gyroscope,
            SAMPLE_DELAY_US
        )

        Log.i(TAG, "Started recording at ${TARGET_SAMPLE_RATE_HZ}Hz")
    }

    /**
     * Stop recording and return the complete IMU recording
     */
    fun stopRecording(): IMURecording {
        sensorManager.unregisterListener(this)
        _isRecording.value = false

        val endTimeMs = System.currentTimeMillis()

        val recording = IMURecording(
            samples = samples.toList(),
            startTimeMs = startTimeMs,
            endTimeMs = endTimeMs,
            sampleRateHz = TARGET_SAMPLE_RATE_HZ,
            deviceModel = deviceModel
        )

        Log.i(TAG, "Stopped. Captured ${samples.size} samples over ${recording.durationMs}ms")
        analyzeRecording(recording)

        return recording
    }

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                // Convert from m/s² to G (divide by 9.81)
                lastAccelX = event.values[0].toDouble() / SensorManager.GRAVITY_EARTH
                lastAccelY = event.values[1].toDouble() / SensorManager.GRAVITY_EARTH
                lastAccelZ = event.values[2].toDouble() / SensorManager.GRAVITY_EARTH
                hasAccelData = true
            }
            Sensor.TYPE_GYROSCOPE -> {
                lastGyroX = event.values[0].toDouble()
                lastGyroY = event.values[1].toDouble()
                lastGyroZ = event.values[2].toDouble()
                hasGyroData = true
            }
        }

        // Only add sample when we have both accel and gyro data
        if (hasAccelData && hasGyroData && _isRecording.value) {
            val sample = IMUSample(
                timestampMs = System.currentTimeMillis(),
                accelerometerX = lastAccelX,
                accelerometerY = lastAccelY,
                accelerometerZ = lastAccelZ,
                gyroscopeX = lastGyroX,
                gyroscopeY = lastGyroY,
                gyroscopeZ = lastGyroZ
            )

            samples.add(sample)
            _sampleCount.value = samples.size
            _currentAcceleration.value = sample.accelerationMagnitude

            hasAccelData = false
            hasGyroData = false
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // Not used but required by interface
    }

    /**
     * Analyze recording for jump detection (local preview)
     */
    private fun analyzeRecording(recording: IMURecording) {
        if (recording.samples.isEmpty()) return

        val accelerations = recording.samples.map { it.accelerationMagnitude }

        val maxAccel = accelerations.maxOrNull() ?: 0.0
        val minAccel = accelerations.minOrNull() ?: 0.0
        val avgAccel = accelerations.average()

        Log.i(TAG, "📊 IMU Analysis:")
        Log.i(TAG, "   Samples: ${recording.sampleCount}")
        Log.i(TAG, "   Duration: ${recording.durationMs}ms")
        Log.i(TAG, "   Max G: ${"%.2f".format(maxAccel)}")
        Log.i(TAG, "   Min G: ${"%.2f".format(minAccel)}")
        Log.i(TAG, "   Avg G: ${"%.2f".format(avgAccel)}")

        // Detect jump phases
        val takeoffIndex = accelerations.indexOfFirst { it > 1.5 }
        if (takeoffIndex >= 0) {
            val takeoffTime = recording.samples[takeoffIndex].timestampMs - recording.startTimeMs
            Log.i(TAG, "   Takeoff detected at: ${takeoffTime}ms")
        }

        val freefallIndex = accelerations.indexOfFirst { it < 0.3 }
        if (freefallIndex >= 0) {
            val freefallTime = recording.samples[freefallIndex].timestampMs - recording.startTimeMs
            Log.i(TAG, "   Freefall detected at: ${freefallTime}ms")
        }

        val landingIndex = accelerations.indexOfLast { it > 2.0 }
        if (landingIndex >= 0) {
            val landingTime = recording.samples[landingIndex].timestampMs - recording.startTimeMs
            Log.i(TAG, "   Landing detected at: ${landingTime}ms")
        }
    }

    /**
     * Serialize IMU recording to JSON
     */
    fun serializeRecording(recording: IMURecording): String {
        return Json.encodeToString(recording)
    }

    /**
     * Create a mock recording for testing
     */
    fun mockRecording(): IMURecording {
        val mockSamples = mutableListOf<IMUSample>()
        val startTime = System.currentTimeMillis()

        // Generate 300 samples (3 seconds at 100Hz)
        for (i in 0 until 300) {
            val t = i / 100.0 // Time in seconds
            val timestampMs = startTime + (i * 10)

            // Simulate jump phases
            val accelMagnitude = when {
                t < 0.5 -> 1.0 + (Math.random() - 0.5) * 0.2 // Standing
                t < 0.7 -> 1.8 + (Math.random() - 0.5) * 0.4 // Takeoff
                t < 1.3 -> 0.1 + (Math.random() - 0.5) * 0.1 // Freefall
                t < 1.5 -> 2.5 + (Math.random() - 0.5) * 0.6 // Landing
                else -> 1.0 + (Math.random() - 0.5) * 0.2 // Recovery
            }

            mockSamples.add(
                IMUSample(
                    timestampMs = timestampMs,
                    accelerometerX = 0.0,
                    accelerometerY = -accelMagnitude,
                    accelerometerZ = 0.0,
                    gyroscopeX = (Math.random() - 0.5) * 0.2,
                    gyroscopeY = (Math.random() - 0.5) * 0.2,
                    gyroscopeZ = (Math.random() - 0.5) * 0.2
                )
            )
        }

        return IMURecording(
            samples = mockSamples,
            startTimeMs = startTime,
            endTimeMs = startTime + 3000,
            sampleRateHz = 100.0,
            deviceModel = deviceModel
        )
    }
}

File: apps/jump-ios/YPJump/xLENS/IMURecorder.swift
// ═══════════════════════════════════════════════════════════════
// IMU RECORDER
// High-frequency motion data capture for xLENS verification
// THE MOAT: AI can fake pixels, not synchronized G-forces
// ═══════════════════════════════════════════════════════════════

import Foundation
import CoreMotion
import Combine

/// Records accelerometer and gyroscope data at ~100Hz
/// This is the unfakeable signal that closes the "Analog Hole"
@MainActor
final class IMURecorder: ObservableObject {

    // MARK: - Published State

    @Published private(set) var isRecording = false
    @Published private(set) var sampleCount = 0
    @Published private(set) var currentAcceleration: Double = 1.0 // In G

    // MARK: - Private Properties

    private let motionManager = CMMotionManager()
    private var samples: [IMUSample] = []
    private var startTime: Int64 = 0
    private var recordingQueue = DispatchQueue(label: "com.ypjump.imu", qos: .userInteractive)

    // Target 100Hz sampling
    private let targetSampleRateHz: Double = 100.0
    private var sampleInterval: TimeInterval { 1.0 / targetSampleRateHz }

    // MARK: - Initialization

    init() {
        // Check device capabilities
        if !motionManager.isAccelerometerAvailable {
            print("⚠️ IMURecorder: Accelerometer not available")
        }
        if !motionManager.isGyroAvailable {
            print("⚠️ IMURecorder: Gyroscope not available")
        }
    }

    // MARK: - Recording Control

    /// Start recording IMU data
    func startRecording() {
        guard !isRecording else { return }
        guard motionManager.isAccelerometerAvailable && motionManager.isGyroAvailable else {
            print("❌ IMURecorder: Motion sensors not available")
            return
        }

        // Reset state
        samples.removeAll()
        sampleCount = 0
        startTime = Int64(Date().timeIntervalSince1970 * 1000)
        isRecording = true

        // Configure sample rates
        motionManager.accelerometerUpdateInterval = sampleInterval
        motionManager.gyroUpdateInterval = sampleInterval

        // Start device motion (combines accelerometer + gyroscope)
        motionManager.deviceMotionUpdateInterval = sampleInterval
        motionManager.startDeviceMotionUpdates(to: .main) { [weak self] motion, error in
            guard let self = self, let motion = motion else { return }

            let now = Int64(Date().timeIntervalSince1970 * 1000)

            // User acceleration (without gravity)
            let userAccel = motion.userAcceleration
            // Gravity vector
            let gravity = motion.gravity
            // Total acceleration = user + gravity
            let totalAccelX = userAccel.x + gravity.x
            let totalAccelY = userAccel.y + gravity.y
            let totalAccelZ = userAccel.z + gravity.z

            // Rotation rate
            let rotation = motion.rotationRate

            let sample = IMUSample(
                timestampMs: now,
                accelerometerX: totalAccelX,
                accelerometerY: totalAccelY,
                accelerometerZ: totalAccelZ,
                gyroscopeX: rotation.x,
                gyroscopeY: rotation.y,
                gyroscopeZ: rotation.z
            )

            Task { @MainActor in
                self.samples.append(sample)
                self.sampleCount = self.samples.count
                self.currentAcceleration = sample.accelerationMagnitude
            }
        }

        print("✅ IMURecorder: Started recording at \(targetSampleRateHz)Hz")
    }

    /// Stop recording and return the complete IMU recording
    func stopRecording() -> IMURecording {
        motionManager.stopDeviceMotionUpdates()
        isRecording = false

        let endTime = Int64(Date().timeIntervalSince1970 * 1000)

        let recording = IMURecording(
            samples: samples,
            startTimeMs: startTime,
            endTimeMs: endTime,
            sampleRateHz: targetSampleRateHz,
            deviceModel: deviceModel
        )

        print("✅ IMURecorder: Stopped. Captured \(samples.count) samples over \(recording.durationMs)ms")

        // Analyze the recording
        analyzeRecording(recording)

        return recording
    }

    // MARK: - Analysis

    /// Analyze recording for jump detection (local preview, server does authoritative analysis)
    private func analyzeRecording(_ recording: IMURecording) {
        guard !recording.samples.isEmpty else { return }

        let accelerations = recording.samples.map { $0.accelerationMagnitude }

        // Find key events
        let maxAccel = accelerations.max() ?? 0
        let minAccel = accelerations.min() ?? 0
        let avgAccel = accelerations.reduce(0, +) / Double(accelerations.count)

        print("📊 IMU Analysis:")
        print("   Samples: \(recording.sampleCount)")
        print("   Duration: \(recording.durationMs)ms")
        print("   Max G: \(String(format: "%.2f", maxAccel))")
        print("   Min G: \(String(format: "%.2f", minAccel))")
        print("   Avg G: \(String(format: "%.2f", avgAccel))")

        // Detect jump phases
        // 1. Takeoff spike (typically >1.5g)
        // 2. Freefall period (~0g)
        // 3. Landing spike (typically >2g)

        if let takeoffIndex = accelerations.firstIndex(where: { $0 > 1.5 }) {
            let takeoffTime = recording.samples[takeoffIndex].timestampMs - recording.startTimeMs
            print("   Takeoff detected at: \(takeoffTime)ms")
        }

        if let freefallIndex = accelerations.firstIndex(where: { $0 < 0.3 }) {
            let freefallTime = recording.samples[freefallIndex].timestampMs - recording.startTimeMs
            print("   Freefall detected at: \(freefallTime)ms")
        }

        if let landingIndex = accelerations.lastIndex(where: { $0 > 2.0 }) {
            let landingTime = recording.samples[landingIndex].timestampMs - recording.startTimeMs
            print("   Landing detected at: \(landingTime)ms")
        }
    }

    // MARK: - Serialization

    /// Serialize IMU recording to JSON data
    func serializeRecording(_ recording: IMURecording) -> Data? {
        let encoder = JSONEncoder()
        encoder.outputFormatting = .sortedKeys // Deterministic output for hashing
        return try? encoder.encode(recording)
    }

    // MARK: - Device Info

    private var deviceModel: String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }
}

// MARK: - Preview Helpers

extension IMURecorder {
    /// Create a mock recording for previews
    static func mockRecording() -> IMURecording {
        var samples: [IMUSample] = []
        let startTime = Int64(Date().timeIntervalSince1970 * 1000)

        // Generate 300 samples (3 seconds at 100Hz)
        for i in 0..<300 {
            let t = Double(i) / 100.0 // Time in seconds
            let timestampMs = startTime + Int64(i * 10)

            // Simulate jump: normal -> takeoff spike -> freefall -> landing spike -> normal
            var accelMagnitude: Double
            if t < 0.5 {
                accelMagnitude = 1.0 + Double.random(in: -0.1...0.1) // Standing
            } else if t < 0.7 {
                accelMagnitude = 1.8 + Double.random(in: -0.2...0.2) // Takeoff
            } else if t < 1.3 {
                accelMagnitude = 0.1 + Double.random(in: -0.05...0.05) // Freefall
            } else if t < 1.5 {
                accelMagnitude = 2.5 + Double.random(in: -0.3...0.3) // Landing
            } else {
                accelMagnitude = 1.0 + Double.random(in: -0.1...0.1) // Recovery
            }

            samples.append(IMUSample(
                timestampMs: timestampMs,
                accelerometerX: 0,
                accelerometerY: -accelMagnitude, // Gravity is down
                accelerometerZ: 0,
                gyroscopeX: Double.random(in: -0.1...0.1),
                gyroscopeY: Double.random(in: -0.1...0.1),
                gyroscopeZ: Double.random(in: -0.1...0.1)
            ))
        }

        return IMURecording(
            samples: samples,
            startTimeMs: startTime,
            endTimeMs: startTime + 3000,
            sampleRateHz: 100,
            deviceModel: "iPhone15,2"
        )
    }
}
